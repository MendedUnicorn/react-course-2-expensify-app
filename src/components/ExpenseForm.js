import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/gm)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide both description and amount',
      }));
    } else {
      this.setState(() => ({
        error: '',
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  //   renderMonthElement = ({ month, onMonthSelect, onYearSelect }) => (
  //     <div style={{ display: 'flex', justifyContent: 'center' }}>
  //       <div>
  //         <select
  //           value={month.month()}
  //           onChange={(e) => onMonthSelect(month, e.target.value)}
  //         >
  //           {moment.months().map((label, value) => (
  //             <option value={value}>{label}</option>
  //           ))}
  //         </select>
  //       </div>
  //       <div>
  //         <select
  //           value={month.year()}
  //           onChange={(e) => onYearSelect(month, e.target.value)}
  //         >
  //           <option value={moment().year() - 1}>Last year</option>
  //           <option value={moment().year()}>{moment().year()}</option>
  //           <option value={moment().year() + 1}>Next year</option>
  //         </select>
  //       </div>
  //     </div>
  //   );

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type='text'
            name=''
            id=''
            autoFocus
            placeholder='Description'
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type='text'
            name=''
            id=''
            placeholder='Amount'
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            // renderMonthElement={this.renderMonthElement}
          />
          <textarea
            name=''
            id=''
            cols='30'
            rows='10'
            placeholder='Add a note for your expense (optional)'
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
