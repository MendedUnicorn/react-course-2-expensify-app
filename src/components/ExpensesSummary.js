import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import expensesSelector from '../selectors/expenses';
import expensesTotalSelector from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format(
    '$ 0,0.00'
  );
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
      </h1>
    </div>
  );
};

const mapStateToProps = (state) => ({
  expensesTotal: expensesTotalSelector(
    expensesSelector(state.expenses, state.filters)
  ),
  expenseCount: expensesSelector(state.expenses, state.filters).length,
});

export default connect(mapStateToProps)(ExpensesSummary);
