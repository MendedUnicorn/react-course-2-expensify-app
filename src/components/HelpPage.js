import React from "react"

const HelpPage = () => <div className=''>This is from my help component</div>;


export default HelpPage