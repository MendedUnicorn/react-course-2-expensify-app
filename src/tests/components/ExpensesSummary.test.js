import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

// let wrapper
// beforeEach(() => {
//     wrapper = shallow(<ExpensesSummary expenses={expenses} />)

// })

test('should render component with no data', () => {
  const wrapper = shallow(<ExpensesSummary />);
  expect(wrapper).toMatchSnapshot();
});

test('should render component with 1 expense', () => {
  const wrapper = shallow(
    <ExpensesSummary expensesTotal={235} expenseCount={1} />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('h1').text()).toBe('Viewing 1 expense totalling $ 2.35');
});

test('should render component with multiple expense', () => {
  const wrapper = shallow(
    <ExpensesSummary expensesTotal={241234} expenseCount={23} />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('h1').text()).toBe(
    'Viewing 23 expenses totalling $ 2,412.34'
  );
});
