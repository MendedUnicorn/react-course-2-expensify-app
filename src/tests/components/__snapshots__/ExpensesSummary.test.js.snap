// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $ 2.35
  </h1>
</div>
`;

exports[`should render component with multiple expense 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $ 2,412.34
  </h1>
</div>
`;

exports[`should render component with no data 1`] = `
<div>
  <h1>
    Viewing 
     
    expenses
     totalling 
    $ 0.00
  </h1>
</div>
`;
